File: kotlin/05-functions/HigherOrder.kt
// Higher-order functions, lambdas, and closures

// Takes a function (Int, Int) -> Int as its third parameter
fun calculate(a: Int, b: Int, operation: (Int, Int) -> Int): Int {
    return operation(a, b)
}

// Returns a function — the returned lambda "closes over" factor
fun makeMultiplier(factor: Int): (Int) -> Int {
    return { value -> value * factor }
}

fun main() {
    // Pass a lambda directly (trailing-lambda syntax)
    val sum = calculate(5, 3) { x, y -> x + y }
    println("Sum: $sum")

    val product = calculate(5, 3) { x, y -> x * y }
    println("Product: $product")

    // Store a function in a variable
    val square: (Int) -> Int = { n -> n * n }
    println("Square of 6: ${square(6)}")

    // A closure that remembers its captured factor
    val triple = makeMultiplier(3)
    println("Triple 7: ${triple(7)}")

    // Built-in higher-order functions on collections
    val numbers = listOf(1, 2, 3, 4, 5)
    val doubled = numbers.map { it * 2 }
    val evens = numbers.filter { it % 2 == 0 }
    println("Doubled: $doubled")
    println("Evens: $evens")
}

File: swift/05-functions/higher_order.swift
// A function that takes another function as a parameter
func applyTwice(_ value: Int, using transform: (Int) -> Int) -> Int {
    return transform(transform(value))
}

// A named function used as an argument
func square(_ x: Int) -> Int {
    return x * x
}

print("applyTwice square 3 = \(applyTwice(3, using: square))")

// A closure stored in a constant
let addOne = { (x: Int) -> Int in x + 1 }
print("addOne(10) = \(addOne(10))")

// Higher-order functions on collections with trailing-closure syntax
let numbers = [1, 2, 3, 4, 5, 6]
let doubled = numbers.map { $0 * 2 }
let evens = numbers.filter { $0 % 2 == 0 }
let sum = numbers.reduce(0, +)

print("Doubled: \(doubled)")
print("Evens: \(evens)")
print("Sum: \(sum)")

// A function that returns a function, capturing `amount`
func makeAdder(_ amount: Int) -> (Int) -> Int {
    return { value in value + amount }
}

let addFive = makeAdder(5)
print("addFive(20) = \(addFive(20))")
